core/src/cluster_nodes.rs

```diff
@@ -0,0 +1,441 @@
+            NodeId::Pubkey(_) => None,
+            NodeId::ContactInfo(node) => Some(node),
+        }
+    }
+}
+
+impl<T> ClusterNodes<T> {
+    pub fn num_peers(&self) -> usize {
+        self.index.len()
+    }
+
+    // A peer is considered live if they generated their contact info recently.
+    pub fn num_peers_live(&self, now: u64) -> usize {
+        self.index
+            .iter()
+            .filter_map(|(_, index)| self.nodes[*index].contact_info())
+            .filter(|node| {
+                let elapsed = if node.wallclock < now {
+                    now - node.wallclock
+                } else {
+                    node.wallclock - now
+                };
+                elapsed < CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS
+            })
+            .count()
+    }
+}
+
+impl ClusterNodes<BroadcastStage> {
+    pub fn new(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Self {
+        new_cluster_nodes(cluster_info, stakes)
+    }
+
+    /// Returns the root of turbine broadcast tree, which the leader sends the
+    /// shred to.
+    pub fn get_broadcast_peer(&self, shred_seed: [u8; 32]) -> Option<&ContactInfo> {
+        if self.index.is_empty() {
+            None
+        } else {
+            let index = weighted_best(&self.index, shred_seed);
+            match &self.nodes[index].node {
+                NodeId::ContactInfo(node) => Some(node),
+                NodeId::Pubkey(_) => panic!("this should not happen!"),
+            }
+        }
+    }
+}
+
+impl ClusterNodes<RetransmitStage> {
+    pub fn new(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Self {
+        new_cluster_nodes(cluster_info, stakes)
+    }
+
+    pub fn get_retransmit_peers(
+        &self,
+        shred_seed: [u8; 32],
+        fanout: usize,
+        slot_leader: Option<Pubkey>,
+    ) -> (
+        Vec<&ContactInfo>, // neighbors
+        Vec<&ContactInfo>, // children
+    ) {
+        // Exclude leader from list of nodes.
+        let index = self.index.iter().copied();
+        let (weights, index): (Vec<u64>, Vec<usize>) = match slot_leader {
+            None => {
+                error!("unknown leader for shred slot");
+                index.unzip()
+            }
+            Some(slot_leader) if slot_leader == self.pubkey => {
+                error!("retransmit from slot leader: {}", slot_leader);
+                index.unzip()
+            }
+            Some(slot_leader) => index
+                .filter(|(_, i)| self.nodes[*i].pubkey() != slot_leader)
+                .unzip(),
+        };
+        let index: Vec<_> = {
+            let shuffle = weighted_shuffle(&weights, shred_seed);
+            shuffle.into_iter().map(|i| index[i]).collect()
+        };
+        let self_index = index
+            .iter()
+            .position(|i| self.nodes[*i].pubkey() == self.pubkey)
+            .unwrap();
+        let (neighbors, children) = compute_retransmit_peers(fanout, self_index, &index);
```

Comment:
Is this filter: https://github.com/solana-labs/solana/blob/e7b36c84841e7a3e8d635bec2e7c0ee18150f212/core/src/retransmit_stage.rs#L443-L447 for `self` no longer necessary on `the neighbors`?

Reply:
That logic, like before, is in `retransmit_stage.rs`, where it says:
```
            // First neighbor is this node itself, so skip it.
            ClusterInfo::retransmit_to(
                &neighbors[1..],
```

Reply:
Ah, missed that bit, thanks!